File: web_extras/routers/scheduler.py
"""HTTP surface for the scheduler subsystem (Wave 97).

Endpoints:

- ``GET    /api/scheduler/schedules``                       list all
- ``POST   /api/scheduler/schedules``                       create
- ``GET    /api/scheduler/schedules/{id}``                  show one
- ``PATCH  /api/scheduler/schedules/{id}``                  update
- ``DELETE /api/scheduler/schedules/{id}``                  remove
- ``POST   /api/scheduler/schedules/{id}/run-now``          fire immediately
- ``GET    /api/scheduler/schedules/{id}/history?limit=20`` recent runs
- ``POST   /api/scheduler/validate-cron``                   validate + preview

The validate-cron endpoint also returns the next 5 firing times for
the supplied expression so the FE can render a "next runs" preview
without round-trips per row.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Body, HTTPException, Query

from backend.core.scheduler import (
    Schedule,
    get_runner,
    get_store,
    next_after,
    validate,
)
from backend.core.scheduler.cron import CronParseError


router = APIRouter(prefix="/api/scheduler", tags=["scheduler"])


# ---------- helpers --------------------------------------------------------


def _ensure_enabled() -> None:
    store = get_store()
    if not store.enabled:
        raise HTTPException(status_code=503, detail="scheduler_store_disabled")


def _schedule_to_dict(s: Schedule) -> dict[str, Any]:
    return s.to_dict()


# ---------- CRUD -----------------------------------------------------------


@router.get("/schedules")
async def list_schedules(
    playbook_id: str | None = Query(default=None),
    only_enabled: bool = Query(default=False),
) -> dict[str, Any]:
    _ensure_enabled()
    store = get_store()
    schedules = await store.list_schedules(
        playbook_id=playbook_id, only_enabled=only_enabled
    )
    return {
        "ok": True,
        "count": len(schedules),
        "schedules": [_schedule_to_dict(s) for s in schedules],
    }


@router.post("/schedules")
async def create_schedule(
    payload: dict[str, Any] = Body(default_factory=dict),
) -> dict[str, Any]:
    _ensure_enabled()
    playbook_id = str(payload.get("playbook_id") or "").strip()
    cron_expression = str(payload.get("cron_expression") or payload.get("cron") or "").strip()
    if not playbook_id:
        raise HTTPException(status_code=400, detail="playbook_id_required")
    if not cron_expression:
        raise HTTPException(status_code=400, detail="cron_expression_required")
    args = payload.get("args") or {}
    if not isinstance(args, dict):
        raise HTTPException(status_code=400, detail="args_must_be_object")
    store = get_store()
    try:
        sched = await store.create_schedule(
            playbook_id=playbook_id,
            cron_expression=cron_expression,
            timezone=str(payload.get("timezone") or "UTC"),
            args=args,
            max_concurrent=int(payload.get("max_concurrent") or 1),
            enabled=bool(payload.get("enabled", True)),
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {"ok": True, "schedule": _schedule_to_dict(sched)}


@router.get("/schedules/{schedule_id}")
async def get_schedule(schedule_id: str) -> dict[str, Any]:
    _ensure_enabled()
    sched = await get_store().get_schedule(schedule_id)
    if sched is None:
        raise HTTPException(status_code=404, detail="schedule_not_found")
    return {"ok": True, "schedule": _schedule_to_dict(sched)}


@router.patch("/schedules/{schedule_id}")
async def patch_schedule(
    schedule_id: str,
    payload: dict[str, Any] = Body(default_factory=dict),
) -> dict[str, Any]:
    _ensure_enabled()
    # Whitelist fields the client may patch.
    updates: dict[str, Any] = {}
    for key in (
        "cron_expression",
        "cron",
        "timezone",
        "enabled",
        "args",
        "max_concurrent",
    ):
        if key in payload:
            target = "cron_expression" if key == "cron" else key
            updates[target] = payload[key]
    try:
        sched = await get_store().update_schedule(schedule_id, updates)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if sched is None:
        raise HTTPException(status_code=404, detail="schedule_not_found")
    return {"ok": True, "schedule": _schedule_to_dict(sched)}


@router.delete("/schedules/{schedule_id}")
async def delete_schedule(schedule_id: str) -> dict[str, Any]:
    _ensure_enabled()
    deleted = await get_store().delete_schedule(schedule_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="schedule_not_found")
    return {"ok": True, "deleted": True}


# ---------- run-now + history ---------------------------------------------


@router.post("/schedules/{schedule_id}/run-now")
async def run_now(schedule_id: str) -> dict[str, Any]:
    _ensure_enabled()
    store = get_store()
    sched = await store.get_schedule(schedule_id)
    if sched is None:
        raise HTTPException(status_code=404, detail="schedule_not_found")
    runner = get_runner()
    # update_next=False so manual runs don't shift the cadence.
    out = await runner.fire_schedule(sched, update_next=False)
    return {"ok": True, "schedule_id": schedule_id, "result": out}


@router.get("/schedules/{schedule_id}/history")
async def history(
    schedule_id: str,
    limit: int = Query(default=20, ge=1, le=200),
) -> dict[str, Any]:
    _ensure_enabled()
    store = get_store()
    sched = await store.get_schedule(schedule_id)
    if sched is None:
        raise HTTPException(status_code=404, detail="schedule_not_found")
    runs = await store.history(schedule_id, limit=limit)
    return {
        "ok": True,
        "schedule_id": schedule_id,
        "count": len(runs),
        "runs": [r.to_dict() for r in runs],
    }


# ---------- cron validation -----------------------------------------------


@router.post("/validate-cron")
async def validate_cron(
    payload: dict[str, Any] = Body(default_factory=dict),
) -> dict[str, Any]:
    """Validate a cron expression and preview the next 5 firing times.

    Body:

    - ``expression`` (required) — the cron expression to test.
    - ``timezone`` (optional, default ``UTC``) — TZ database name.
    """

    expression = str(payload.get("expression") or "").strip()
    tz_name = str(payload.get("timezone") or "UTC").strip() or "UTC"
    if not expression:
        return {"valid": False, "error": "expression_required"}
    if not validate(expression):
        return {"valid": False, "error": "invalid_cron_expression"}
    # Compute next 5 fires walking forward.
    next_runs: list[str] = []
    cursor = datetime.now(timezone.utc)
    try:
        for _ in range(5):
            nxt = next_after(expression, cursor, tz=tz_name)
            next_runs.append(nxt.isoformat())
            cursor = nxt
    except CronParseError as exc:
        return {"valid": False, "error": str(exc)}
    return {
        "valid": True,
        "expression": expression,
        "timezone": tz_name,
        "next_5_runs": next_runs,
    }

File: experiments/neural-showcase-v3/src/lib/scheduler.ts
// SYNC: claude-w97-scheduler
/**
 * scheduler.ts — thin REST client for the Wave 97 scheduler engine.
 *
 * All endpoints live under /api/scheduler/*. The cron engine is a
 * pure-stdlib parser on the backend (no croniter dep), so the
 * `validateCron` helper round-trips to it for both validation and
 * the next-5-runs preview.
 */

import { API_BASE } from "@/lib/api";

export interface Schedule {
  id: string;
  playbook_id: string;
  cron_expression: string;
  timezone: string;
  enabled: boolean;
  last_run_at: number | null;
  next_run_at: number | null;
  last_status: string | null;
  max_concurrent: number;
  args: Record<string, unknown>;
  created_at: number;
}

export interface RunRecord {
  id: string;
  schedule_id: string;
  started_at: number;
  finished_at: number | null;
  status: string;
  output_summary: string | null;
  trace_id: string | null;
  duration_ms: number | null;
}

export interface CronValidation {
  valid: boolean;
  expression?: string;
  timezone?: string;
  next_5_runs?: string[];
  error?: string;
}

/** POST /api/scheduler/validate-cron — server-side validation + preview. */
export async function validateCron(
  expression: string,
  timezone: string = "UTC",
): Promise<CronValidation> {
  try {
    const r = await fetch(`${API_BASE}/api/scheduler/validate-cron`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ expression, timezone }),
    });
    if (!r.ok) return { valid: false, error: `HTTP ${r.status}` };
    return (await r.json()) as CronValidation;
  } catch (e) {
    return { valid: false, error: (e as Error).message };
  }
}

/** GET /api/scheduler/schedules */
export async function listSchedules(playbookId?: string): Promise<Schedule[]> {
  const qs = playbookId
    ? `?playbook_id=${encodeURIComponent(playbookId)}`
    : "";
  const r = await fetch(`${API_BASE}/api/scheduler/schedules${qs}`);
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const body = (await r.json()) as { schedules?: Schedule[] };
  return body.schedules ?? [];
}

/** POST /api/scheduler/schedules */
export async function createSchedule(payload: {
  playbook_id: string;
  cron_expression: string;
  timezone?: string;
  args?: Record<string, unknown>;
  max_concurrent?: number;
  enabled?: boolean;
}): Promise<Schedule> {
  const r = await fetch(`${API_BASE}/api/scheduler/schedules`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const body = (await r.json()) as { schedule: Schedule };
  return body.schedule;
}

/** PATCH /api/scheduler/schedules/{id} */
export async function patchSchedule(
  id: string,
  updates: Partial<{
    cron_expression: string;
    timezone: string;
    enabled: boolean;
    args: Record<string, unknown>;
    max_concurrent: number;
  }>,
): Promise<Schedule> {
  const r = await fetch(
    `${API_BASE}/api/scheduler/schedules/${encodeURIComponent(id)}`,
    {
      method: "PATCH",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(updates),
    },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const body = (await r.json()) as { schedule: Schedule };
  return body.schedule;
}

/** DELETE /api/scheduler/schedules/{id} */
export async function deleteSchedule(id: string): Promise<void> {
  const r = await fetch(
    `${API_BASE}/api/scheduler/schedules/${encodeURIComponent(id)}`,
    { method: "DELETE" },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
}

/** POST /api/scheduler/schedules/{id}/run-now */
export async function runScheduleNow(id: string): Promise<unknown> {
  const r = await fetch(
    `${API_BASE}/api/scheduler/schedules/${encodeURIComponent(id)}/run-now`,
    { method: "POST" },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return (await r.json()) as unknown;
}

/** GET /api/scheduler/schedules/{id}/history?limit=N */
export async function fetchHistory(
  id: string,
  limit: number = 20,
): Promise<RunRecord[]> {
  const r = await fetch(
    `${API_BASE}/api/scheduler/schedules/${encodeURIComponent(id)}/history?limit=${limit}`,
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const body = (await r.json()) as { runs?: RunRecord[] };
  return body.runs ?? [];
}

/** Convenience: POST /api/playbooks/{id}/schedule */
export async function schedulePlaybook(
  playbookId: string,
  payload: {
    cron: string;
    timezone?: string;
    args?: Record<string, unknown>;
    enabled?: boolean;
  },
): Promise<Schedule> {
  const r = await fetch(
    `${API_BASE}/api/playbooks/${encodeURIComponent(playbookId)}/schedule`,
    {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const body = (await r.json()) as { schedule: Schedule };
  return body.schedule;
}
